### PR TITLE
Listing content on /advertising by number of views - POC 1

### PR DESCRIPTION
**Pull request by paulopmt1**
Related to 1693-gh-Automattic/dotcom-forge

## Proposed Changes

* Here we're showing /advertising posts/pages by: 
  - 5 most recent posts
  - up to 20 most viewed
  - up to 20 most commented (if doesn't find 10 most viewed)
  - up to 20 products

## Testing Instructions

* Apply the Diff in your sandbox: https://github.com/Automattic/jetpack/pull/29149
* Access a popular site you have and make sure we're listing posts and pages by their views at the top of /advertising items
* Tip: Comment [these lines](https://github.com/Automattic/wp-calypso/blob/update/filter-advertising-by-views-v2/client/my-sites/promote-post/main.tsx#L123-L144) and access any private site from automattic that have more views, like the explorers p2.
* Test around with Atomic (will need to sync Jetpack repo to a WoA blog) and Simple sites
* Make sure it works with sites without any views (in this case we'll load pages and posts by comments):

image

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by candy02058912** [comment]
Linking the thread on what's causing the infinite loop here for visibility: p1677235259718559/1677235241.569579-slack-CKZHG0QCR

File: client/my-sites/promote-post/main.tsx [hunk 1]
```diff
@@ -81,6 +94,21 @@ export default function PromotedPosts( { tab }: Props ) {
...
 	);
 
+	const topViewedPostAndPages = useSelector( ( state ) =>
+		getTopPostAndPages( state, selectedSiteId, topPostsQuery )
+	);
+
+	const mostPopularPostAndPages = useSelector( ( state ) => {
+		const topPosts: any[ PostType ] = [];
+
+		topViewedPostAndPages?.map( ( post: any ) => {
```
Comment (candy02058912) [inline client/my-sites/promote-post/main.tsx]: `map` is mostly used when we need to return an array; in this case, `forEach` might be a better option

File: client/my-sites/promote-post/main.tsx [hunk 2]
```diff
@@ -81,6 +94,21 @@ export default function PromotedPosts( { tab }: Props ) {
...
+			const item = getSitePost( state, selectedSiteId, post.id );
+			item && topPosts.push( { ...item, views: post.views } );
+		} );
+
+		return topPosts;
```
Comment (candy02058912) [inline client/my-sites/promote-post/main.tsx]: Should we also return the pages?

Comment (paulopmt1) [reply client/my-sites/promote-post/main.tsx]: Good point, will add the parameter to return `any` type of posts.

File: client/my-sites/promote-post/main.tsx (hunk 1 above)
Comment (paulopmt1) [reply client/my-sites/promote-post/main.tsx]: Nice, [Improved it here](https://github.com/Automattic/wp-calypso/pull/73760/commits/520a576acf5ba98efa46ef9880e227110fe24e6f).

File: client/my-sites/promote-post/main.tsx (hunk 2 above)
Comment (paulopmt1) [reply client/my-sites/promote-post/main.tsx]: Noted it also returns pages, so [I just renamed the variable here](https://github.com/Automattic/wp-calypso/pull/73760/commits/905023029e6a132ce88d9a7290707a23fa6334f5).

**Review by lupus2k — APPROVED** [review]
Good job, @paulopmt1 & @candy02058912, it works perfect 🎉

File: client/my-sites/promote-post/main.tsx [hunk 3]
```diff
@@ -141,9 +184,16 @@ export default function PromotedPosts( { tab }: Props ) {
...
 	}
 
-	const content = [ ...( posts || [] ), ...( pages || [] ), ...( products || [] ) ];
-
-	const isLoading = isLoadingPage && isLoadingPost && isLoadingProducts;
+	const content = [
+		...( mostPopularPostAndPages || [] ),
+		...( postAndPagesByComments || [] ),
+		...( products || [] ),
+	];
```
Comment (DustyReagan) [inline client/my-sites/promote-post/main.tsx]: Do we need to de-dupe these somehow? I managed to see the same post twice.

de-dupping

Comment (paulopmt1) [reply client/my-sites/promote-post/main.tsx]: > Do we need to de-dupe these somehow?

Good catch @DustyReagan, [Fixed it here](https://github.com/Automattic/wp-calypso/pull/73760/commits/8d184c2ed26828d644e69509f115c23854b5b1fa).

Comment (DustyReagan) [reply client/my-sites/promote-post/main.tsx]: Appears to deduping posts now! Nice!

**Review by DustyReagan — APPROVED** [review]
LGTM! Nice work! 👍

**Comment by paulopmt1** [comment]
Thanks for the reviews folks!

Waiting for [the Jetpack PR to be deployed](https://github.com/Automattic/jetpack/pull/29149) so I'll deploy this one.

**Comment by enejb** [comment]
Nice improvement! 

I tested this PR using a User Agent that that matches the Jetpack Apps ( `Mozilla/5.0 (iPhone; CPU iPhone OS 14_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Mobile/15E148 wp-iphone` ) 

And I didn't see any regressions.

**Pull request merged**